src/kvstore/wal/FileBasedWal.cpp

```diff
@@ -449,32 +403,46 @@ bool FileBasedWal::appendLogInternal(BufferPtr& buffer,
         return false;
     }
 
-    if (buffer &&
-        (buffer->size() +
-         msg.size() +
-         sizeof(ClusterID) +
-         sizeof(TermID) +
-         sizeof(LogID) > maxBufferSize_)) {
-        // Freeze the current buffer
-        if (buffer->freeze()) {
-            flusher_->flushBuffer(shared_from_this(), buffer);
-        }
-        buffer.reset();
-    }
+    // Write to the WAL file first
+    std::string strBuf;
+    strBuf.reserve(sizeof(LogID)
+                    + sizeof(TermID)
+                    + sizeof(ClusterID)
+                    + msg.size()
+                    + 2 * sizeof(int32_t));
+    strBuf.append(reinterpret_cast<char*>(&id), sizeof(LogID));
+    strBuf.append(reinterpret_cast<char*>(&term), sizeof(TermID));
+    int32_t len = msg.size();
+    strBuf.append(reinterpret_cast<char*>(&len), sizeof(int32_t));
+    strBuf.append(reinterpret_cast<char*>(&cluster), sizeof(ClusterID));
+    strBuf.append(reinterpret_cast<const char*>(msg.data()), msg.size());
+    strBuf.append(reinterpret_cast<char*>(&len), sizeof(int32_t));
+
+    // Prepare the WAL file if it's not opened
+    if (currFd_ < 0) {
+        prepareNewFile(id);
+    } else if (currInfo_->size() + strBuf.size() > maxFileSize_) {
+        // Need to roll over
+        closeCurrFile();
 
-    // Create a new buffer if needed
-    if (!buffer) {
-        std::unique_lock<std::mutex> guard(buffersMutex_);
-        buffer = createNewBuffer(id, guard);
+        std::lock_guard<std::mutex> g(walFilesMutex_);
+        prepareNewFile(id);
```

Comment:
if here need a lock to protected the current file, I think we should protected the file every where we use it.

Reply:
The lock actually protected the walFiles_ (list of WAL info), not the file itself

Reply:
because I find that there is reopen current file in prepareNewFile and rollbackToLog, so if we just lock here, current thread unlock walFilesMutex_ and begin write data to currFd_, then the other thread call rollbackToLog and reopen current file, maybe there is some problem.

Reply:
Good point

We don't expect ```rollbackToLog()``` and ```appendLog()``` to be called simultaneously

Reply:
so maybe we can design it lock free?